Assigning to sumtype results in wrong C codegen

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
OS: linux, Ubuntu 20.04.3 LTS (WSL 2)
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i5-10210U CPU @ 1.60GHz
CC version: cc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0

getwd: /home/aprokurov
vmodules: /home/aprokurov/.vmodules
vroot: /home/aprokurov/v
vexe: /home/aprokurov/v/v
vexe mtime: 2022-01-18 08:05:52
is vroot writable: true
is vmodules writable: true
V full version: V 0.2.4 d826317

Git version: git version 2.25.1
Git vroot status: weekly.2022.03
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 3654d6a8

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg file.v`
```v
pub type Foo = Bar | Baz 


fn (mut f Foo) set_baz() {
	f = Baz{}
}

pub struct Bar {}
pub struct Baz {}

fn main() {
	mut x := Foo(Bar{})
	x.set_baz()
}

```

**What did you expect to see?**

Successful compilation


**What did you see instead?**
```
==================
/tmp/v_1000/file.17017172462154197226.tmp.c:11113: error: cannot convert 'struct main__Foo *' to 'struct main__Foo'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -435,7 +435,12 @@ fn (mut g Gen) gen_assign_stmt(node ast.AssignStmt) {
 						if op_overloaded {
 							g.op_arg(val, op_expected_right, val_type)
 						} else {
-							g.expr_with_cast(val, val_type, var_type)
+							exp_type := if left.is_auto_deref_var() {
+								var_type.deref()
+							} else {
+								var_type
+							}
+							g.expr_with_cast(val, val_type, exp_type)
 						}
 					}
 				}

--- a/vlib/v/tests/mut_receiver_of_sumtype_test.v
+++ b/vlib/v/tests/mut_receiver_of_sumtype_test.v
@@ -1,0 +1,17 @@
+pub type Foo = Bar | Baz
+
+fn (mut f Foo) set_baz() {
+	f = Baz{}
+}
+
+pub struct Bar {}
+
+pub struct Baz {}
+
+fn test_mut_receiver_of_sumtype() {
+	mut x := Foo(Bar{})
+	x.set_baz()
+
+	println(x)
+	assert '$x' == 'Foo(Baz{})'
+}

